src/dmd/s2ir.d

```diff
@@ -184,6 +184,31 @@ private extern (C++) class S2irVisitor : Visitor
         block *bexit = mystate.breakBlock ? mystate.breakBlock : dmd.backend.global.block_calloc();
 
         incUsage(irs, s.loc);
+
+        /* inline.d changes expressions such as `_ctfe ? a : b;` to
+         *      if (__ctfe)
+         *          a;
+         *      else
+         *          b;
+         * Since __ctfe is always false at runtime, the if body can be omitted,
+         * and the statement rewritten simply as `b;`.
+         *
+         * This is necessary for expressions such as `a ~= b`, where `b` is not
+         * an array. In expressionsem.d, they are lowered to:
+         * `__ctfe ? a ~= b : _d_arrayappendcTX(a, 1), a[$ - 1] = b, a;`
+         * If CTFE hasn't already interpreted `a ~= b`, then this branch must be
+         * removed, as e2ir.d no longer supports the runtime hook to
+         * `_d_arrayappendcTX`.
```

Comment:
Can't you just remove it during semantic?

Reply:
I'm not sure I understand what you mean. I create the lowering during semantic3 and the lowering is needed for CTFE. I don't see where in semantic i should remove it.

Reply:
Sounds like something has gone wrong to me.  Lowering should only occur _after_ CTFE has had a chance to reduce the code, not before.

Reply:
The problem is that when the compiler analyzes the body of a function it cannot know if the function is going to be interpreted or not in the future. Since we cannot interpret `_d_arrayappend` this trick with `__ctfe` is used.

Reply:
Performing the lowering in e2ir.d, where it previously was, has the following disadvantages:
1. Maintainability decreases as LDC and GDC will have to implement the lowering themselves.
2. Readability also decreases, because the code in e2ir.d is less expressive than that in expressionsem.d.

While it is true that lowering the `CatAssign` as early as expressionsem.d makes the logic more complex, requiring that e2ir.d and s2ir.d omit the `true` branch of the `__ctfe` cond, I believe this approached is preferable because it elegantly solves the problems above, while also clearly separating the expressions intended for CTFE and for code generation.

Reply:
> I believe this approach is preferable because it elegantly solves the problems above, while also clearly separating the expression intended for CTFE from those meant for code generation.

No it doesn't, you're adding semantic logic to the code generator.  The code generator should at best be just a trivial translation from one representation to another.

What I see is that you've added some new `assert(0)` branches in the code generator. This means that the front-end should _never_ leak these AST constructs to the back-end, and yet, that is exactly what the lowering does, so the code generator is becomes even more complex than it needs to be by having a need to filter AST nodes that should have not have reached it in the first place.

Reply:
Hmm, maybe we can make the hook interpret-able. That would get rid of the filtering code in `e2ir.d` and would still enable us to perform the lowering in the frontend. Is that an acceptable solution @ibuclaw ?